<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Li Lin" />

<meta name="date" content="2021-01-02" />

<title>epiConv Package</title>


<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(title);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">epiConv Package</h1>
<h4 class="author">Li Lin</h4>
<h4 class="date">2021-01-02</h4>



<p>EpiConv is an algorithm of clustering scATAC-seq data and correcting batch effects.<br> To run the scripts below, package <code>SingleCellExperiment</code> is required to store the data. It can be installed through Bioconductor.</p>
<div id="integrate-scatac-seq-data" class="section level3">
<h3>Integrate scATAC-seq data</h3>
<p>We will show the usage of epiConv on LPS-stimulated human bone marrow mononuclear cells (BMMCs, data from <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE123580">GSE123580</a>). The data contains resting and LPS-stimulated BMMCs.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" title="1"><span class="kw">suppressMessages</span>(<span class="kw">library</span>(epiConv))</a>
<a class="sourceLine" id="cb1-2" title="2"><span class="kw">suppressMessages</span>(<span class="kw">library</span>(SingleCellExperiment))</a>
<a class="sourceLine" id="cb1-3" title="3"></a>
<a class="sourceLine" id="cb1-4" title="4">mat&lt;-<span class="kw">readMM</span>(<span class="dt">file=</span><span class="st">&quot;bmmc_matrix.mtx&quot;</span>)</a>
<a class="sourceLine" id="cb1-5" title="5">cell_info&lt;-<span class="kw">read.table</span>(<span class="dt">file=</span><span class="st">&quot;bmmc_ident.tsv&quot;</span>)</a>
<a class="sourceLine" id="cb1-6" title="6"><span class="kw">colnames</span>(cell_info)&lt;-<span class="kw">c</span>(<span class="st">&quot;barcode&quot;</span>,<span class="st">&quot;ident&quot;</span>)</a>
<a class="sourceLine" id="cb1-7" title="7">peak&lt;-<span class="kw">read.table</span>(<span class="dt">file=</span><span class="st">&quot;peaks.bed&quot;</span>)</a>
<a class="sourceLine" id="cb1-8" title="8"><span class="kw">colnames</span>(peak)&lt;-<span class="kw">c</span>(<span class="st">&quot;seqnames&quot;</span>,<span class="st">&quot;start&quot;</span>,<span class="st">&quot;end&quot;</span>)</a>
<a class="sourceLine" id="cb1-9" title="9"><span class="kw">rownames</span>(mat)&lt;-<span class="kw">paste</span>(peak<span class="op">$</span>seqnames,<span class="st">&quot;:&quot;</span>,</a>
<a class="sourceLine" id="cb1-10" title="10">                     peak<span class="op">$</span>start,<span class="st">&quot;-&quot;</span>,</a>
<a class="sourceLine" id="cb1-11" title="11">                     peak<span class="op">$</span>end,<span class="dt">sep=</span><span class="st">&quot;&quot;</span>)</a>
<a class="sourceLine" id="cb1-12" title="12"><span class="kw">colnames</span>(mat)&lt;-cell_info[,<span class="dv">1</span>]</a>
<a class="sourceLine" id="cb1-13" title="13">sce&lt;-<span class="kw">SingleCellExperiment</span>(<span class="dt">assays=</span><span class="kw">SimpleList</span>(<span class="dt">counts=</span>mat),</a>
<a class="sourceLine" id="cb1-14" title="14">                          <span class="dt">rowRanges=</span><span class="kw">GRanges</span>(peak),<span class="dt">colData=</span>cell_info)</a>
<a class="sourceLine" id="cb1-15" title="15"><span class="kw">colData</span>(sce)<span class="op">$</span>lib_size&lt;-<span class="kw">lib.estimate</span>(<span class="kw">assays</span>(sce)<span class="op">$</span>counts)</a>
<a class="sourceLine" id="cb1-16" title="16">sce&lt;-<span class="kw">subset</span>(sce,<span class="dt">select=</span><span class="kw">colData</span>(sce)<span class="op">$</span>lib_size<span class="op">&gt;</span><span class="dv">500</span>)</a>
<a class="sourceLine" id="cb1-17" title="17">sce&lt;-<span class="kw">subset</span>(sce,<span class="dt">subset=</span><span class="kw">freq.estimate</span>(<span class="kw">assays</span>(sce)<span class="op">$</span>counts)<span class="op">&gt;</span><span class="dv">0</span>)</a></code></pre></div>
<p>First we read the data and create a SingleCellExperiment object to store the data.<code>lib.estimate</code> and <code>freq.estimate</code> are two simple functions to calculate the library size of single cells and the frequency of peaks. We removed cells with library size &lt;= 500.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" title="1">Smat&lt;-<span class="kw">run.epiConv</span>(<span class="dt">mat=</span><span class="kw">assays</span>(sce)<span class="op">$</span>counts,</a>
<a class="sourceLine" id="cb2-2" title="2">                  <span class="dt">lib_size=</span>sce<span class="op">$</span>lib_size,</a>
<a class="sourceLine" id="cb2-3" title="3">                  <span class="dt">nbootstrap=</span><span class="dv">15</span>,</a>
<a class="sourceLine" id="cb2-4" title="4">                  <span class="dt">nsample=</span><span class="kw">floor</span>(<span class="kw">nrow</span>(sce)<span class="op">*</span><span class="fl">0.2</span>),</a>
<a class="sourceLine" id="cb2-5" title="5">                  <span class="dt">bin=</span><span class="dv">5000</span>,</a>
<a class="sourceLine" id="cb2-6" title="6">                  <span class="dt">inf_replace=</span>(<span class="op">-</span><span class="dv">8</span>))</a></code></pre></div>
<p>We calculate the similarity matrix of cells by <code>run.epiConv</code>.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" title="1"><span class="kw">reducedDims</span>(sce)<span class="op">$</span>umap&lt;-<span class="kw">run.umap_louvain</span>(<span class="dt">Smat=</span>Smat,</a>
<a class="sourceLine" id="cb3-2" title="2">                                        <span class="dt">knn=</span><span class="dv">50</span>,</a>
<a class="sourceLine" id="cb3-3" title="3">                                        <span class="dt">umap_settings=</span>umap<span class="op">::</span>umap.defaults,</a>
<a class="sourceLine" id="cb3-4" title="4">                                        <span class="dt">resolution=</span><span class="ot">NULL</span>)</a></code></pre></div>
<p>We perform dimension reduction by <code>run.umap_louvain</code>.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" title="1">Smat&lt;-Smat<span class="op">-</span><span class="kw">median</span>(Smat)</a>
<a class="sourceLine" id="cb4-2" title="2">eigs&lt;-<span class="kw">dim.reduce</span>(Smat,<span class="dt">neigs=</span><span class="dv">30</span>)</a>
<a class="sourceLine" id="cb4-3" title="3">residual_mat&lt;-Smat<span class="op">-</span><span class="kw">tcrossprod</span>(eigs<span class="op">$</span>vectors,<span class="kw">t</span>(<span class="kw">t</span>(eigs<span class="op">$</span>vectors)<span class="op">*</span>eigs<span class="op">$</span>values))</a></code></pre></div>
<p>We perform Eigenvalue Decompostion on similarity matrix by <code>dim.reduce</code>.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" title="1">batch&lt;-<span class="kw">factor</span>(sce<span class="op">$</span>ident)</a>
<a class="sourceLine" id="cb5-2" title="2">guide_features&lt;-<span class="kw">sapply</span>(<span class="kw">levels</span>(batch),<span class="cf">function</span>(x){</a>
<a class="sourceLine" id="cb5-3" title="3">  index&lt;-<span class="kw">which</span>(batch<span class="op">==</span>x)</a>
<a class="sourceLine" id="cb5-4" title="4">  eigs<span class="op">$</span>vectors[index,]</a>
<a class="sourceLine" id="cb5-5" title="5">})</a>
<a class="sourceLine" id="cb5-6" title="6">knn_update&lt;-<span class="kw">eigs.knn</span>(<span class="dt">Smat=</span>Smat,</a>
<a class="sourceLine" id="cb5-7" title="7">                     <span class="dt">features=</span>guide_features,</a>
<a class="sourceLine" id="cb5-8" title="8">                     <span class="dt">batch=</span>batch,</a>
<a class="sourceLine" id="cb5-9" title="9">                     <span class="dt">reference=</span><span class="st">&quot;Resting&quot;</span>,</a>
<a class="sourceLine" id="cb5-10" title="10">                     <span class="dt">knn_target=</span><span class="dv">50</span>,</a>
<a class="sourceLine" id="cb5-11" title="11">                     <span class="dt">knn_reference=</span><span class="dv">10</span>,</a>
<a class="sourceLine" id="cb5-12" title="12">                     <span class="dt">threshold=</span><span class="dv">2</span>)</a></code></pre></div>
<p>We create the knn graph bewteen batches and SNN graph within batches to find recurrent cell populations across batches by <code>eigs.knn</code>.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" title="1">eigs_corrected&lt;-<span class="kw">eigs.correct</span>(<span class="dt">eigs=</span>eigs<span class="op">$</span>vectors,</a>
<a class="sourceLine" id="cb6-2" title="2">                             <span class="dt">knn_mat=</span>knn_update,</a>
<a class="sourceLine" id="cb6-3" title="3">                             <span class="dt">batch=</span>batch,</a>
<a class="sourceLine" id="cb6-4" title="4">                             <span class="dt">reference=</span><span class="st">&quot;Resting&quot;</span>,</a>
<a class="sourceLine" id="cb6-5" title="5">                             <span class="dt">knn_transfer_correction=</span><span class="dv">10</span>)</a></code></pre></div>
<p>We scale the Eigen vectors to remove batch effects by <code>eigs.correct</code>.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" title="1">Smat_corrected&lt;-<span class="kw">tcrossprod</span>(eigs_corrected,<span class="kw">t</span>(<span class="kw">t</span>(eigs_corrected)<span class="op">*</span>eigs<span class="op">$</span>values))<span class="op">+</span>residual_mat</a>
<a class="sourceLine" id="cb7-2" title="2">temp&lt;-<span class="kw">run.umap_louvain</span>(<span class="dt">Smat=</span>Smat_corrected,</a>
<a class="sourceLine" id="cb7-3" title="3">                       <span class="dt">knn=</span><span class="dv">50</span>,</a>
<a class="sourceLine" id="cb7-4" title="4">                       <span class="dt">umap_settings=</span>umap<span class="op">::</span>umap.defaults,</a>
<a class="sourceLine" id="cb7-5" title="5">                       <span class="dt">resolution=</span><span class="fl">0.8</span>)</a>
<a class="sourceLine" id="cb7-6" title="6"><span class="kw">reducedDims</span>(sce)<span class="op">$</span>umap_corrected&lt;-<span class="kw">as.matrix</span>(temp[,<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>])</a>
<a class="sourceLine" id="cb7-7" title="7"><span class="kw">colData</span>(sce)&lt;-<span class="kw">cbind</span>(<span class="kw">colData</span>(sce),<span class="dt">cluster=</span><span class="kw">factor</span>(temp[,<span class="dv">3</span>]))</a>
<a class="sourceLine" id="cb7-8" title="8"></a>
<a class="sourceLine" id="cb7-9" title="9"><span class="kw">plot</span>(<span class="kw">reducedDims</span>(sce)<span class="op">$</span>umap,<span class="dt">pch=</span><span class="st">&quot;+&quot;</span>,<span class="dt">cex=</span><span class="fl">0.5</span>,<span class="dt">col=</span><span class="kw">factor</span>(sce<span class="op">$</span>ident))</a>
<a class="sourceLine" id="cb7-10" title="10"><span class="kw">plot</span>(<span class="kw">reducedDims</span>(sce)<span class="op">$</span>umap_corrected,<span class="dt">pch=</span><span class="st">&quot;+&quot;</span>,<span class="dt">cex=</span><span class="fl">0.5</span>,<span class="dt">col=</span><span class="kw">factor</span>(sce<span class="op">$</span>ident))</a>
<a class="sourceLine" id="cb7-11" title="11"><span class="kw">plot</span>(<span class="kw">reducedDims</span>(sce)<span class="op">$</span>umap_corrected,<span class="dt">pch=</span><span class="st">&quot;+&quot;</span>,<span class="dt">cex=</span><span class="fl">0.5</span>,<span class="dt">col=</span><span class="kw">rainbow</span>(<span class="kw">nlevels</span>(sce<span class="op">$</span>cluster))[sce<span class="op">$</span>cluster])</a></code></pre></div>
<p>We calculate the similarity matrix from corrected Eigen vectors and then perfrom dimension reduction and clustering.<br></p>
</div>
<div id="integrate-multi-omics-data-with-scatac-seq-data" class="section level3">
<h3>Integrate multi-omics data with scATAC-seq data</h3>
<p>If cells are with transcriptome profiles (co-assay data), we can use the transcriptome to improve the performance of batch correction. Next, we will integrate one co-assay data of postnatal day 0 mouse cortex (GSE126074) and one scATAC-seq data from <a href="https://support.10xgenomics.com/single-cell-atac/datasets">10x Genomics</a> to show the usage of epiConv. The data after basic quality control is stored in <code>sce.rds</code>.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" title="1"><span class="kw">suppressMessages</span>(<span class="kw">library</span>(epiConv))</a>
<a class="sourceLine" id="cb8-2" title="2"><span class="kw">suppressMessages</span>(<span class="kw">library</span>(SingleCellExperiment))</a>
<a class="sourceLine" id="cb8-3" title="3"></a>
<a class="sourceLine" id="cb8-4" title="4">sce&lt;-<span class="kw">readRDS</span>(<span class="dt">file=</span><span class="st">&quot;sce.rds&quot;</span>)</a>
<a class="sourceLine" id="cb8-5" title="5">index_coassay&lt;-<span class="kw">which</span>(<span class="kw">colData</span>(sce)<span class="op">$</span>ident<span class="op">==</span><span class="st">&quot;co-assay&quot;</span>)</a>
<a class="sourceLine" id="cb8-6" title="6">Smat&lt;-<span class="kw">run.epiConv</span>(<span class="dt">mat=</span><span class="kw">assays</span>(sce)<span class="op">$</span>counts[,index_coassay],</a>
<a class="sourceLine" id="cb8-7" title="7">                  <span class="dt">lib_size=</span>sce<span class="op">$</span>lib_size[index_coassay],</a>
<a class="sourceLine" id="cb8-8" title="8">                  <span class="dt">nbootstrap=</span><span class="dv">15</span>,</a>
<a class="sourceLine" id="cb8-9" title="9">                  <span class="dt">nsample=</span><span class="kw">floor</span>(<span class="kw">nrow</span>(sce)<span class="op">*</span><span class="fl">0.2</span>),</a>
<a class="sourceLine" id="cb8-10" title="10">                  <span class="dt">bin=</span><span class="dv">5000</span>,</a>
<a class="sourceLine" id="cb8-11" title="11">                  <span class="dt">inf_replace=</span>(<span class="op">-</span><span class="dv">8</span>))</a>
<a class="sourceLine" id="cb8-12" title="12"></a>
<a class="sourceLine" id="cb8-13" title="13">feature_coassay&lt;-<span class="kw">cal.feature</span>(<span class="dt">Smat=</span>Smat,</a>
<a class="sourceLine" id="cb8-14" title="14">                             <span class="dt">pcs=</span><span class="kw">reducedDims</span>(sce)<span class="op">$</span>PCA_RNA[index_coassay,],</a>
<a class="sourceLine" id="cb8-15" title="15">                             <span class="dt">neigs=</span><span class="dv">50</span>)</a></code></pre></div>
<p>First, we need to combine RNA-seq and ATAC-seq data to resolve the relationships of single cells in co-assay data. We combined 50 principal components (PCs) from RNA-seq <code>reducedDims(sce)$PCA_RNA</code>(calculated from RNA-seq) and 50 Eigen vectors from ATAC-seq by <code>cal.feature</code>. <code>feature_coassay</code> contains 100 features in total and will be used in following steps.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" title="1">Smat&lt;-<span class="kw">run.epiConv</span>(<span class="dt">mat=</span><span class="kw">assays</span>(sce)<span class="op">$</span>counts,</a>
<a class="sourceLine" id="cb9-2" title="2">                  <span class="dt">lib_size=</span>sce<span class="op">$</span>lib_size,</a>
<a class="sourceLine" id="cb9-3" title="3">                  <span class="dt">nbootstrap=</span><span class="dv">15</span>,</a>
<a class="sourceLine" id="cb9-4" title="4">                  <span class="dt">nsample=</span><span class="kw">floor</span>(<span class="kw">nrow</span>(sce)<span class="op">*</span><span class="fl">0.2</span>),</a>
<a class="sourceLine" id="cb9-5" title="5">                  <span class="dt">bin=</span><span class="dv">10000</span>,</a>
<a class="sourceLine" id="cb9-6" title="6">                  <span class="dt">inf_replace=</span>(<span class="op">-</span><span class="dv">8</span>))</a>
<a class="sourceLine" id="cb9-7" title="7"><span class="kw">reducedDims</span>(sce)<span class="op">$</span>umap&lt;-<span class="kw">run.umap_louvain</span>(<span class="dt">Smat=</span>Smat,</a>
<a class="sourceLine" id="cb9-8" title="8">                                        <span class="dt">knn=</span><span class="ot">NULL</span>,</a>
<a class="sourceLine" id="cb9-9" title="9">                                        <span class="dt">umap_settings=</span><span class="ot">NULL</span>,</a>
<a class="sourceLine" id="cb9-10" title="10">                                        <span class="dt">resolution=</span><span class="ot">NULL</span>)</a>
<a class="sourceLine" id="cb9-11" title="11">Smat&lt;-Smat<span class="op">-</span><span class="kw">median</span>(Smat)</a>
<a class="sourceLine" id="cb9-12" title="12">eigs&lt;-<span class="kw">dim.reduce</span>(Smat,<span class="dt">neigs=</span><span class="dv">30</span>)</a>
<a class="sourceLine" id="cb9-13" title="13">residual_mat&lt;-Smat<span class="op">-</span><span class="kw">tcrossprod</span>(eigs<span class="op">$</span>vectors,<span class="kw">t</span>(<span class="kw">t</span>(eigs<span class="op">$</span>vectors)<span class="op">*</span>eigs<span class="op">$</span>values))</a>
<a class="sourceLine" id="cb9-14" title="14"></a>
<a class="sourceLine" id="cb9-15" title="15">batch&lt;-<span class="kw">factor</span>(<span class="kw">colData</span>(sce)<span class="op">$</span>ident)</a>
<a class="sourceLine" id="cb9-16" title="16">guide_features&lt;-<span class="kw">sapply</span>(<span class="kw">levels</span>(batch),<span class="cf">function</span>(x){</a>
<a class="sourceLine" id="cb9-17" title="17">  index&lt;-<span class="kw">which</span>(batch<span class="op">==</span>x)</a>
<a class="sourceLine" id="cb9-18" title="18">  eigs<span class="op">$</span>vectors[index,]</a>
<a class="sourceLine" id="cb9-19" title="19">})</a></code></pre></div>
<p>We calculate the similarity matrix on combined data and perfrom Eigenvalue Decomposition.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" title="1">guide_features[[<span class="st">&quot;co-assay&quot;</span>]]&lt;-<span class="kw">cbind</span>(guide_features[[<span class="st">&quot;co-assay&quot;</span>]],</a>
<a class="sourceLine" id="cb10-2" title="2">                                    <span class="kw">reducedDims</span>(sce)<span class="op">$</span>PCA_RNA[index_coassay,])</a></code></pre></div>
<p>We add Principal Components (PCs) from RNA-seq to guide features. PCs from RNA-seq improve the results of knn matrix when the sequencing depth of ATAC-seq is shallow.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" title="1">knn_update&lt;-<span class="kw">eigs.knn</span>(<span class="dt">Smat=</span>Smat,</a>
<a class="sourceLine" id="cb11-2" title="2">                     <span class="dt">features=</span>guide_features,</a>
<a class="sourceLine" id="cb11-3" title="3">                     <span class="dt">batch=</span>batch,</a>
<a class="sourceLine" id="cb11-4" title="4">                     <span class="dt">reference=</span><span class="st">&quot;ATAC&quot;</span>,</a>
<a class="sourceLine" id="cb11-5" title="5">                     <span class="dt">knn_target=</span><span class="dv">50</span>,</a>
<a class="sourceLine" id="cb11-6" title="6">                     <span class="dt">knn_reference=</span><span class="dv">10</span>,</a>
<a class="sourceLine" id="cb11-7" title="7">                     <span class="dt">threshold=</span><span class="dv">2</span>)</a></code></pre></div>
<p>We create the knn graph.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" title="1">knn_update[[<span class="st">&quot;co-assay&quot;</span>]][[<span class="st">&quot;co-assay&quot;</span>]]&lt;-<span class="kw">cal.snn</span>(<span class="dt">Smat=</span><span class="kw">as.matrix</span>(<span class="kw">dist</span>(feature_coassay))<span class="op">*</span>(<span class="op">-</span><span class="dv">1</span>),</a>
<a class="sourceLine" id="cb12-2" title="2">                                                <span class="dt">knn=</span><span class="kw">floor</span>(<span class="kw">nrow</span>(feature_coassay)<span class="op">*</span><span class="fl">0.01</span>))</a></code></pre></div>
<p>The SNN graph within batches are calculated based on ATAC-seq only. If RNA-seq data is available, we can calculate the SNN graph from RNA-seq and ATAC-seq by <code>cal.snn</code> instead to get better results.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" title="1">eigs_corrected&lt;-<span class="kw">eigs.correct</span>(<span class="dt">eigs=</span>eigs<span class="op">$</span>vectors,</a>
<a class="sourceLine" id="cb13-2" title="2">                             <span class="dt">knn_mat=</span>knn_update,</a>
<a class="sourceLine" id="cb13-3" title="3">                             <span class="dt">batch=</span>batch,</a>
<a class="sourceLine" id="cb13-4" title="4">                             <span class="dt">reference=</span><span class="st">&quot;ATAC&quot;</span>,</a>
<a class="sourceLine" id="cb13-5" title="5">                             <span class="dt">knn_transfer_correction=</span><span class="dv">10</span>)</a>
<a class="sourceLine" id="cb13-6" title="6"></a>
<a class="sourceLine" id="cb13-7" title="7">Smat_corrected&lt;-<span class="kw">tcrossprod</span>(eigs_corrected,<span class="kw">t</span>(<span class="kw">t</span>(eigs_corrected)<span class="op">*</span>eigs<span class="op">$</span>values))<span class="op">+</span>residual_mat</a>
<a class="sourceLine" id="cb13-8" title="8"></a>
<a class="sourceLine" id="cb13-9" title="9">temp&lt;-<span class="kw">run.umap_louvain</span>(<span class="dt">Smat=</span>Smat_corrected,</a>
<a class="sourceLine" id="cb13-10" title="10">                       <span class="dt">knn=</span><span class="ot">NULL</span>,</a>
<a class="sourceLine" id="cb13-11" title="11">                       <span class="dt">umap_settings=</span><span class="ot">NULL</span>,</a>
<a class="sourceLine" id="cb13-12" title="12">                       <span class="dt">resolution=</span><span class="kw">c</span>(<span class="fl">0.4</span>,<span class="fl">0.8</span>)) <span class="co">##Here we try resolution=0.4 and 0.8.</span></a>
<a class="sourceLine" id="cb13-13" title="13"><span class="kw">reducedDims</span>(sce)<span class="op">$</span>umap_corrected&lt;-<span class="kw">as.matrix</span>(temp[,<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>])</a>
<a class="sourceLine" id="cb13-14" title="14"><span class="kw">colData</span>(sce)&lt;-<span class="kw">cbind</span>(<span class="kw">colData</span>(sce),<span class="dt">cluster=</span><span class="kw">factor</span>(temp[,<span class="dv">3</span>]))</a>
<a class="sourceLine" id="cb13-15" title="15"><span class="kw">plot</span>(<span class="kw">reducedDims</span>(sce)<span class="op">$</span>umap,<span class="dt">pch=</span><span class="st">&quot;+&quot;</span>,<span class="dt">cex=</span><span class="fl">0.5</span>,<span class="dt">col=</span>batch)</a>
<a class="sourceLine" id="cb13-16" title="16"><span class="kw">plot</span>(<span class="kw">reducedDims</span>(sce)<span class="op">$</span>umap_corrected,<span class="dt">pch=</span><span class="st">&quot;+&quot;</span>,<span class="dt">cex=</span><span class="fl">0.5</span>,<span class="dt">col=</span>batch)</a>
<a class="sourceLine" id="cb13-17" title="17"><span class="kw">plot</span>(<span class="kw">reducedDims</span>(sce)<span class="op">$</span>umap_corrected,<span class="dt">pch=</span><span class="st">&quot;+&quot;</span>,<span class="dt">cex=</span><span class="fl">0.5</span>,</a>
<a class="sourceLine" id="cb13-18" title="18">     <span class="dt">col=</span><span class="kw">rainbow</span>(<span class="kw">nlevels</span>(sce<span class="op">$</span>cluster))[sce<span class="op">$</span>cluster])</a></code></pre></div>
<p>Finally we see the results.</p>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
